### PR TITLE
mgr/dashboard: rgw policy group management api

### PR DESCRIPTION
**Pull request by nizamial09**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by Pegonzal — COMMENTED** [review]
LGTM @nizamial09, once the PR is out of draft if there are no big changes it gets my approval.

I left a bunch of comments, but very minor things.
A part from those, should the `SYNC INFO` be added as well? [DOC](https://docs.ceph.com/en/latest/radosgw/multisite-sync-policy/#sync-info)

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 1]
```diff
@@ -1744,3 +1756,156 @@ def get_primary_zonedata(self, data):
...
+            raise DashboardException(error, http_status_code=500, component='rgw')
+
+    def get_sync_policy_group(self, group_id: str, bucket_name: str = ''):
+        rgw_sync_policy_cmd = ['sync', 'group', 'get', '--group-id', group_id]
+        if bucket_name:
+            rgw_sync_policy_cmd += ['--bucket', bucket_name]
+        try:
+            exit_code, out, _ = mgr.send_rgwadmin_command(rgw_sync_policy_cmd)
+            if exit_code > 0:
+                raise DashboardException('Unable to get sync policy',
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/services/rgw_client.py]: done

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 2]
```diff
@@ -114,15 +116,87 @@ def migrate(self, daemon_name=None, realm_name=None, zonegroup_name=None, zone_n
...
+    @DeletePermission
+    def remove_sync_policy_group(self, group_id: str, bucket_name=''):
+        multisite_instance = RgwMultisite()
+        return multisite_instance.remove_sync_policy_group(group_id, bucket_name)
+
+    @Endpoint(method='POST', path='/sync_flow')
+    @EndpointDoc("Create the sync flow")
+    @CreatePermission
+    def create_sync_flow(self, flow_id: str, flow_type: SyncFlowTypes, group_id: str,
+                         source_zone='', destination_zone='', zones='', bucket_name=''):
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/rgw.py]: thanks

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 3]
```diff
@@ -984,6 +985,17 @@ def perform_validations(self, retention_period_days, retention_period_years, mod
         return retention_period_days, retention_period_years
 
 
+class SyncStatus(Enum):
+    enabled = 'enabled'
+    allowed = 'allowed'
+    forbidden = 'forbidden'
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/services/rgw_client.py:991]: even after this, i had to use the `.value` to get the str value of the Enum. I tried sublcassing the __str__ but that also didn't work, so left it as it is for now.

File: src/pybind/mgr/dashboard/services/rgw_client.py [hunk 4]
```diff
@@ -1744,3 +1756,156 @@ def get_primary_zonedata(self, data):
...
+                raise DashboardException('Unable to remove sync flow',
+                                         http_status_code=500, component='rgw')
+        except SubprocessError as error:
+            raise DashboardException(error, http_status_code=500, component='rgw')
+
+    def create_sync_pipe(self, group_id: str, pipe_id: str, source_zones: List[str] = [],
+                         destination_zones: List[str] = [], bucket_name: str = ''):
+        rgw_sync_policy_cmd = ['sync', 'group', 'pipe', 'create',
+                               '--group-id', group_id, '--pipe-id', pipe_id]
+
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/services/rgw_client.py:1873]: Yup, it was intentional. For now, I just went with the parameters that are required for the UI's basic functionality. As we move towards the advanced functionalities like filtering buckets in pipe, we can extend this method with those options. IMO, that way we can also test those things individually.

**Comment by nizamial09** [comment]
@Pegonzal thanks for the reviews. Addressed most of them and left some comments inline for the things i didn't address.

File: src/pybind/mgr/dashboard/services/rgw_client.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/services/rgw_client.py:991]: Oh, my bad, I was convinced otherwise

**Comment by nizamial09** [comment]
> A part from those, should the `SYNC INFO` be added as well? [DOC](https://docs.ceph.com/en/latest/radosgw/multisite-sync-policy/#sync-info)

not now as the sync info is a bit hard to understand as it is. I heard it might get refactored later on. So probably we can check later.

File: src/pybind/mgr/dashboard/controllers/rgw.py [hunk 5]
```diff
@@ -114,15 +115,94 @@ def migrate(self, daemon_name=None, realm_name=None, zonegroup_name=None, zone_n
...
         multisite_instance = RgwMultisite()
         result = multisite_instance.get_multisite_sync_status()
         return result
 
+    @Endpoint(path='/sync_policy')
+    @EndpointDoc("Get the sync policy")
+    @ReadPermission
+    def get_sync_policy(self, bucket_name=''):
+        multisite_instance = RgwMultisite()
+        return multisite_instance.get_sync_policy(bucket_name)
```
[the remark this answers is not in the document]
Comment (nizamial09) [reply src/pybind/mgr/dashboard/controllers/rgw.py]: thanks, addressed

**Comment by nizamial09** [comment]
> however, they do not require period update.

I agree, they do not require period update. But period update is necessary when different zones are coming into picture and we need the current policies changes to be carried over to the other zones as well. For now I removed the period commit from this API because its a time consuming process and ideally the group creation action itself shouldn't be failed on waiting for the period commit to be done. I'll think about it and add it separately.

**Comment by nizamial09** [comment]
> can we just updated conditionally like so?:

I am going to isolate the period commit into a separate task rather than executing along with these actions because it can cause timeouts. I'll take it up in a second PR.

**Review by cloudbehl — APPROVED** [review]

**Pull request merged**